llama-index-integrations/graph_stores/llama-index-graph-stores-nebula/llama_index/graph_stores/nebula/nebula_property_graph.py

```diff
@@ -0,0 +1,932 @@
+
+    @property
+    def client(self):
+        """client of NebulaGraph."""
+        return self._client
+
+    def _execute(self, query: str) -> ResultSet:
+        return self._client.execute(query)
+
+    def refresh_schema(self) -> None:
+        """
+        Example data of self.structured_schema:
+        {
+            "node_props": {
+                "Person": [
+                    {"property": "name", "type": "STRING", "comment": "The name of the person"},
+                    {"property": "age", "type": "INTEGER", "comment": "The age of the person"},
+                    {"property": "dob", "type": "DATE", "comment": "The date of birth of the person"}
+                ],
+                "Company": [
+                    {"property": "name", "type": "STRING", "comment": "The name of the company"},
+                    {"property": "founded", "type": "DATE", "comment": "The date of foundation of the company"}
+                ]
+            },
+            "rel_props": {
+                "WORKS_AT": [
+                    {"property": "since", "type": "DATE", "comment": "The date when the person started working at the company"}
+                ],
+                "MANAGES": [
+                    {"property": "since", "type": "DATE", "comment": "The date when the person started managing the company"}
+                ]
+            },
+            "relationships": [
+                {"start": "Person", "type": "WORKS_AT", "end": "Company"},
+                {"start": "Person", "type": "MANAGES", "end": "Company"}
+            ]
+        }
+        """
+
+        tags_schema = {}
+        edge_types_schema = {}
+        relationships = []
+        for node_label in self.structured_query(
+            "MATCH ()-[node_label:`__meta__node_label__`]->() "
+            "RETURN node_label.label AS name, "
+            "JSON_EXTRACT(node_label.props_json) AS props"
+        ):
+            tags_schema[node_label["name"]] = []
+            # TODO: add properties to tags_schema
+        for rel_label in self.structured_query(
+            "MATCH ()-[rel_label:`__meta__rel_label__`]->() "
+            "RETURN rel_label.label AS name, "
+            "src(rel_label) AS src, dst(rel_label) AS dst, "
+            "JSON_EXTRACT(rel_label.props_json) AS props"
+        ):
+            edge_types_schema[rel_label["name"]] = []
+            # TODO: add properties to edge_types_schema
+            relationships.append(
+                {
+                    "start": rel_label["src"],
+                    "type": rel_label["name"],
+                    "end": rel_label["dst"],
+                }
+            )
+
+        self.structured_schema = {
+            "node_props": tags_schema,
+            "rel_props": edge_types_schema,
+            "relationships": relationships,
+            # TODO: need to check necessarity of meta data here
+        }
+
+        # Update node info
+        # TODO need to look into this enhanced schema
```

Comment:
Yea the "enhanced schema" is something that neo4j wrote. Since neo4j doesn't have an actual schema, they needed to write some giant function. I think it was mostly copied from langchain 😅

Reply:
Aha, got it @logan-markewich , thanks, we'll remove this.

Also regarding text2cypher, we'll need further things to make it work with NebulaGraph(non-llm dialect&& schema self-reflection correctors)

Reply:
Yes! There is an actually an attribute on the base class of the graph store for the text2cypher prompt. We can maybe override it in the nebula class so a more specific one is used?
https://github.com/run-llama/llama_index/blob/957a0b98b4ef5a8305ba09c17340c39c2f22616c/llama-index-core/llama_index/core/graph_stores/types.py#L257

The text2cypher retriever also has an option to pass in some cypher-cleaner callable, if nebula has something specific for this
https://github.com/run-llama/llama_index/blob/957a0b98b4ef5a8305ba09c17340c39c2f22616c/llama-index-core/llama_index/core/indices/property_graph/sub_retrievers/text_to_cypher.py#L43